File: src/main/java/me/yoursole/ctf/Events/PlayerDeath.java
package me.yoursole.ctf.Events;

import me.yoursole.ctf.DataFiles.GameData;
import me.yoursole.ctf.DataFiles.Items.Flag;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.PlayerRespawnEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

import java.util.ArrayList;
import java.util.Objects;

public class PlayerDeath implements Listener {
    private static Player a= null;
    @EventHandler
    public void onPlayerDeath(PlayerDeathEvent e){
        if(GameData.it==null)
            return;
        if(Objects.equals(Objects.requireNonNull(e.getEntity().getPlayer()).getUniqueId(), GameData.it.getUniqueId())){
            e.getEntity().getInventory().remove(Flag.flag);
            e.getEntity().getPlayer().setGlowing(false);

            e.getDrops().clear();
            boolean found = false;
            for(Player player : Bukkit.getOnlinePlayers()){
                if (player != e.getEntity().getPlayer()) {
                    if(Objects.requireNonNull(e.getDeathMessage()).contains(player.getDisplayName())){
                        found = true;
                        player.getInventory().addItem(Flag.flag);
                        player.sendMessage(ChatColor.GREEN+"You received the flag through murder");
                        player.setAbsorptionAmount(player.getAbsorptionAmount() + 5);
                        player.addPotionEffect(GameData.regen);
                        player.setGlowing(true);
                        for(Player playera : Bukkit.getOnlinePlayers()){
                            if(!player.getDisplayName().equals(playera.getDisplayName())){
                                playera.sendMessage(ChatColor.GREEN+player.getDisplayName()+" has received the flag");
                            }
                        }
                        GameData.it=player;
                        break;
                    }
                }
            }
            if(!found){
                if(Bukkit.getOnlinePlayers().size()>1){
                    Player player = Bukkit.getOnlinePlayers().stream().skip((int) (Bukkit.getOnlinePlayers().size() * Math.random())).findFirst().orElse(null);
                    while(Objects.equals(player, e.getEntity().getPlayer())){
                        player = Bukkit.getOnlinePlayers().stream().skip((int) (Bukkit.getOnlinePlayers().size() * Math.random())).findFirst().orElse(null);
                    }
                    assert player != null;
                    player.getInventory().addItem(Flag.flag);
                    player.setAbsorptionAmount(player.getAbsorptionAmount() + 5);
                    player.addPotionEffect(GameData.regen);
                    player.sendMessage(ChatColor.GREEN+"You got lucky and received the flag");
                    player.setGlowing(true);
                    if(GameData.inNether&&!GameData.netherHunters.contains(player)){
                        GameData.inNether=false;
                    }
                    for(Player playera : Bukkit.getOnlinePlayers()){
                        if(!player.getDisplayName().equals(playera.getDisplayName())){
                            playera.sendMessage(ChatColor.GREEN+player.getDisplayName()+" has received the flag");
                        }
                    }
                    GameData.it=player;
                }else{
                    a=e.getEntity().getPlayer();
                }



            }
        }else{
            GameData.netherHunters.remove(e.getEntity().getPlayer());
        }


    }
    @EventHandler
    public void onRespawn(PlayerRespawnEvent e){
        if(e.getPlayer().equals(a)){
            Objects.requireNonNull(e.getPlayer()).getInventory().addItem(Flag.flag);
            e.getPlayer().setGlowing(true);
            e.getPlayer().setAbsorptionAmount(5);
            e.getPlayer().sendMessage(ChatColor.GREEN+"You got lucky...you are the only person online so you get to keep the flag");
            a=null;
        }
    }
}

File: src/main/kotlin/me/yoursole/ctf/events/PlayerDeath.kt
package me.yoursole.ctf.events

import me.yoursole.ctf.datafiles.FlagDropper
import me.yoursole.ctf.datafiles.GameData
import me.yoursole.ctf.datafiles.items.Flag
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.PlayerRespawnEvent

object PlayerDeath : Listener {
    @EventHandler
    fun onPlayerDeath(e: PlayerDeathEvent) {
        if (GameData.it == null) return
        if (e.player.uniqueId == GameData.it!!.uniqueId) {
            e.player.inventory.remove(Flag.flag)
            e.player.isGlowing = false
            e.drops.clear()
            val killer = e.player.killer ?: Bukkit.getOnlinePlayers()
                .find {
                    it.gameMode == GameMode.SURVIVAL && it.uniqueId != GameData.it?.uniqueId && e.deathMessage!!.contains(
                        it.displayName
                    )
                }
            if (killer != null && killer.uniqueId != e.player.uniqueId) {
                killer.inventory.addItem(Flag.flag)
                killer.sendMessage("§aYou received the flag through murder")
                killer.absorptionAmount += 5
                killer.addPotionEffect(GameData.regen)
                killer.isGlowing = true
                for (player in Bukkit.getOnlinePlayers()) {
                    if (killer.displayName != player.displayName) {
                        player.sendMessage("§a${killer.displayName} has received the flag")
                    }
                }
                GameData.it = killer
            } else {
                if (GameData.inNether) {
                    val chosen = Bukkit.getOnlinePlayers().filter { it.gameMode == GameMode.SURVIVAL && it != GameData.it }
                        .minByOrNull { it.location.distanceSquared(e.entity.location) }
                    if (chosen != null) {
                        chosen.inventory.addItem(Flag.flag)
                        chosen.absorptionAmount += 5
                        chosen.addPotionEffect(GameData.regen)
                        chosen.sendMessage("§aYou were the closest to the player, you have received the flag.")
                        chosen.isGlowing = true
                        if (GameData.inNether && !GameData.netherHunters.contains(chosen)) {
                            GameData.inNether = false
                        }
                        for (playera in Bukkit.getOnlinePlayers()) {
                            if (chosen.displayName != playera.displayName) {
                                playera.sendMessage("§a${chosen.displayName} has received the flag")
                            }
                        }
                        GameData.it = chosen
                    } else {
                        a = e.player
                    }
                } else {
                    GameData.it = null
                    GameData.itLoc = null
                    e.player.inventory.remove(Flag.flag)
                    e.player.isGlowing = false
                    e.drops.clear()
                    FlagDropper.dropFlag(GameData.world!!, e.player.location.blockX, e.player.location.blockZ, 30)
                }
            }
        } else {
            GameData.netherHunters.remove(e.player)
            if (e.player != GameData.it && e.player.killer?.uniqueId == GameData.it!!.uniqueId) {
                GameData.scores.compute(GameData.it!!.uniqueId) { _, i ->
                    (i ?: 0) + 10
                }
            }
        }
    }

    @EventHandler
    fun onRespawn(e: PlayerRespawnEvent) {
        if (e.player == a) {
            (e.player).inventory.addItem(Flag.flag)
            e.player.isGlowing = true
            e.player.absorptionAmount = 5.0
            e.player.sendMessage("§aYou got lucky...you are the only person online so you get to keep the flag")
            a = null
        }
    }

    private var a: Player? = null
}